lang_python.adoc

```diff
@@ -33,68 +33,94 @@ Runtime library API version must match compiler version that you're using
 (i.e. if you use ksc v0.8, then you must use compatible v0.8 runtime).
 
 [[add-runtime-release]]
-==== Using release version
+==== Using release version (NOT recommended)
 
 With stable version of Kaitai Struct compiler, one can use use following
 methods:
 
 * On Debian / Ubuntu, one can install it from apt repositories:
 
-** For Python 2.x:
+** For Python 3.x:
 +
 [source,shell]
-sudo apt-get install python-kaitaistruct
+sudo apt-get install python3-kaitaistruct
 +
-Reference: https://packages.debian.org/buster/python-kaitaistruct[package info for Debian], https://packages.ubuntu.com/eoan/python-kaitaistruct[package info for Ubuntu]
+Reference: https://packages.debian.org/buster/python3-kaitaistruct[package info for Debian], https://packages.ubuntu.com/eoan/python3-kaitaistruct[package info for Ubuntu]
 
-** For Python 3.x:
+** For Python 2.x (DEPRECATED, python 2 is EOL since 2020-01-01):
 +
 [source,shell]
-sudo apt-get install python3-kaitaistruct
+sudo apt-get install python-kaitaistruct
 +
-Reference: https://packages.debian.org/buster/python3-kaitaistruct[package info for Debian], https://packages.ubuntu.com/eoan/python3-kaitaistruct[package info for Ubuntu]
+Reference: https://packages.debian.org/buster/python-kaitaistruct[package info for Debian], https://packages.ubuntu.com/eoan/python-kaitaistruct[package info for Ubuntu]
+
 
-* If your project is using *requirements.txt* to manage your
-  dependencies, add the following line to it:
+* Add the following dependency specifier into
+`install_requires` of `options` section
+of your `setup.cfg` (or the arg of `setuptools.setup`
+with the same name in your `setup.py`) :
 +
 ----
 kaitaistruct
 ----
 +
-and then run `pip install -r requirements.txt` to update all your
-dependencies.
+and then run `pip install --upgrade -e ./` in the dir with the `setup.py`
```

Comment:
>Please use python3 -m pip ... instead of pip ...: https://snarky.ca/why-you-should-use-python-m-pip/

IMHO, one should just do

`bash
git clone --depth=1 https://github.com/pypa/pip.git
cd pip
<needed_python_interpreter> ./setup.py bdist_wheel
<needed_python_interpreter> -m pip install --upgrade ./dist/*.whl
```,

this will make the needed pip the interpreter available without `<needed_python_interpreter> -m`.

Reply:
I'm not sure how these steps are supposed to help here? The main reason why you're supposed to use `python -m pip` instead of `pip` is that there's no guarantee that `pip` refers to the same Python interpreter as `python`. There are other reasons too, which are explained in the blog post that I've linked. The steps that you've suggested don't help against any of those problems.

Besides, installing things directly from Git for no reason is a terrible idea. You have no idea what state the repo is currently in - it's most likely a development version that could have all sorts of bugs, incomplete features or breaking changes. You also have to run setup.py, which you almost never have to do when installing from PyPI, because every decent package provides pre-built wheels. Also, any depenencies of the project will by default be installed from PyPI anyway, which makes the whole process kind of pointless.

Reply:
>The main reason why you're supposed to use python -m pip instead of pip is that there's no guarantee that pip refers to the same Python interpreter as python.

When you install pip yourself you control which interpreter will be used - the one you use to install `pip` itself.

>Besides, installing things directly from Git for no reason is a terrible idea.

IMHO it is a very good idea.

* On git is the live code resides, that is supposed to be seen by the ones developing it. On pypi there are zip archives noone looks at. 

* One gets the latest features and bug fixes. I have been used PEP 508 in my packages a half of a year before this feature has landed to a release.

* >You have no idea what state the repo is currently in - it's most likely a development version that could have all sorts of bugs, incomplete features or breaking changes.

Then I want to encounter them before they are standardized as features. It is always YOUR FAULT if YOUR PACKAGE is incompatible with the latest version of its dependency. One day that version of the will land and your users will have to downgrade it becaus of your package. The general rule: you can always expect people to upgrade the software when the upgrade is free (not requiring buying anything or rewriting anything from scratch or doing some costly research or doing some costly computations, I hope you understand me), but you cannot expect people to downgrade software. `venvs` are not a solution, they are just shit some lazy people use hoping that they will help them not to upgrade their code.

>You also have to run setup.py, which you almost never have to do when installing from PyPI, because every decent package provides pre-built wheels.

For pip it is feasible. For numpy it is semi-feasible. For tensorflow or plaidml it is completely infeasible.

>Also, any depenencies of the project will by default be installed from PyPI anyway, which makes the whole process kind of pointless.

It is disableable.

Reply:
> When you install pip yourself you control which interpreter will be used - the one you use to install `pip` itself.

That's not how this works. Installing `pip` (or any other package) does not change your PATH. If the `pip` command refers to a different interpreter than you need/expect, then reinstalling the `pip` package changes nothing about that - the `pip` command will still refer to exactly the same (wrong) interpreter as before.

> IMHO it is a very good idea.

OK, if that workflow works for you, great - I'm not stopping you from installing packages this way on your system. But the usual way to install dependencies in Python is by having your dependency management tool download pre-packaged release versions from PyPI. The KS Python guide should follow that established standard, unless there's a very good reason why KS specifically needs a different workflow (which is not the case as far as I can tell).